### PR TITLE
Generic TWI/I2C interface for Marlin

### PR DESCRIPTION
**Pull request by jbrazio**
This PR started as a feature request at #2895, the current status is EXPERIMENTAL, to enable it one must `#define EXPERIMENTAL_I2CBUS` (~~this was intentionally left undocumented in `Configuration.h` so it's easy for developers to test the feature but hard enough for users to break stuff~~).

### CODE REVIEW

**Comment by thinkyhead** [comment]
I like this general idea, and I was also thinking of the option of having multiple arguments with the same letter. Of course it extends some "tenets" of GCode, but I'm not _altogether_ against that. Some of the "rules of GCode" that Marlin has already "broken" include things like having bare string arguments (it is, after all, _Numerical_ control), re-using letter codes with different meanings, using the same letter codes differently (must `V` always be `float`? and `F` always mean feedrate?), etc.

There's a technical reason that we haven't yet had any commands that take the same letter code multiple times. Part of the reason is that _some GCode parsers_ pre-parse all the arguments in a command first before calling the handler function, so each letter argument only gets a single slot and only the last value is kept. I only mention this as a limitation of "reference" parser code which describes how you "should" parse GCode. But it also happens to be the way that more than one firmware currently works.

Obviously, Marlin (coming from Tonokip) can have its own handlers refer right back to the original GCode arguments string and do whatever they like (and we do). But in that regard, we should be moving in the other direction. Namely, making the handlers more agnostic about the source of the numbers they're given. Even to the point of not caring whether `G1` comes at the front or the end of the line.

Philosophically, I agree that the limitations of legacy CNC "GCode" should not be maintained as the limitations of RepRap firmware forever-more. The question only remains, how far to stretch it and in what ways?

Some extensions that we might consider adding to Marlin:
- Option to use delimited lists. Why specify `B` over and over when you could just do this:
  - `M777 S12,345,88,13,0` ?
  - `M777 S12 345 88 13 0` ?
  - `M777 S(12 354 -76.25 98)` ?
  - `G1 (0 0 0)` ?
- Quoted strings support…
  - `M117 "Print \"jarlid\" completed"`
  - `M23 "/Long Dir Name/MyObject.g"`
- Substitutions for strings…
  - `M117 S "Print time $TIME"` … with `S` flag for "display only for SD-printing"
  - `M117 S "Print \"$SDFILE\" completed"`
  - `M117 S "Currently at $SDPROGRESS%"`

Consider the "binary gcode" [S3G Protocol](https://github.com/makerbot/s3g/blob/master/doc/s3gProtocol.md) that ReplicatorG uses, for example. It eschews the whole CNC legacy for a more efficient (thus faster, cooler, less current draw) almost minimal protocol. We might consider supporting S3G later, once Marlin reaches a point where the code is modular enough to plug in different parsers.

Support for more parsers in Marlin might not be too far off, actually. We just need to pre-parse the command arguments before calling into the "GCode" handler function. Then the handlers grab the arguments they need "off the shelf." In fact, I have in mind a hybrid of the parser in Matt Roberts' RepRap Firmware and Marlin that will reduce overhead, add flexibility, and avoid parsing or storing `float` (until a handler actually _asks_ for a `float`).

I'm also thinking along the lines of making the parser more of a data structure, which can really help to optimize it. In the process the data structure would end up being the "complete rules" of Marlin's GCode implementation, and that would be very nice to have in hand.

**Comment by Roxy-3D** [comment]
> There's a technical reason that we haven't yet had any commands that take the same letter code multiple times. ...
> Obviously, Marlin (coming from Tonokip) can have its own handlers refer right back to the original GCode arguments string and do whatever they like (and we do). ...
> Philosophically, I agree that the limitations of legacy CNC "GCode" should not be maintained ...

LOL.....

GCode sucks!!!!

**Comment by CONSULitAS** [comment]
@thinkyhead
I like your approach to pass multiple values as an array.
To simplify the parser, it would be best to chose a magic token starting and ending the reading of the array or string.

I would suggest
- for values
  - `M777 S[1,2,3,4,5,6,7,8,9]`-> `[]` tells parser to fill an array of values (like in c or OpenSCAD)
- for strings
  - `M117 ("Print \"jarlid\" completed")` --> `()` tells parser to handle (and concatenate) quoted strings
  - `M117 S ("Print time ",$TIME)` --> `$` tells parser to substitute with env variable
  - `M117 S ($TIME,\n)` --> `\` tells parser to substitute with non printable char
  - `M117 S ("Print \"",$SDFILE,"\" completed",13)` --> Number (0-255) tells parser to generate char

so the parser could look like

```
if ( currentchar=="[" )
  parsearray(currentpos+1);
else if ( currentchar=='(' )
  parsestring(currentpos+1);
```

**Comment by jbrazio** [comment]
We need to simplify the existing parser, that is currently one of the bottlenecks of Marlin.
Having said that, I would prefer to evolve Marlin into a binary, byte efficient stream like S3G.

@CONSULitAS stuff like this `M777 S[1,2,3,4,5,6,7,8,9]` would only increase parser complexity (if we don't have enough of it already). You're using two **more** bytes as `M777 S1S2S3S4S5S6S7S8S9` and you can't parse it as simple as:

``` cpp
while (*cmd) {
  switch (*cmd) {
    case 'S':
      // do stuff
      break;

    default:
      cmd++;
  }
}
```

Which will allow you to do `M777 S1` as well as `M777 S1 S2`as well as `M777 S1S2` without using spaghetti code.

**Comment by thinkyhead** [comment]
Adding more string-scanning of the original GCode into the handlers is not a positive way to go. Also I think it's a bad idea to support new types of values like arrays, which this feature doesn't require. Altering GCode so it takes the same letter argument multiple times is also unlikely to fly.

Consider [the sad tale of `M160`](http://reprap.org/wiki/G-code#M160:_Number_of_mixed_materials). It extended the `G1` command to use colons to separate multiple values given with the `E` parameter. Someone proposed this method, but not a single firmware implements it. When it came time for pros like Repetier to implement color-mixing, they decided to go with the more granular (but much simpler to parse) `M563`, `M567`, and `M568` that don't require any special parsing. (These are the codes that Marlin will end up using for color mixing too.)

This is a nice PR, so I hate to just throw it out. But, there's no compelling reason to extend the GCode parser (or to put mini-parsers inside handlers) for this feature. So let's try to do a "pure" implementation first.

**Comment by jbrazio** [comment]
I'm not pushing nothing outside the envelope here, in fact this is intended to be used like this as you suggested:

```
; "t2" test macro
; This macro send the string "Marlin"
; It uses multiple M155 commands with one B<base 10> arg
M155 A63  ; Target slave address
M155 B77  ; M
M155 B97  ; a
M155 B114 ; r
M155 B108 ; l
M155 B105 ; i
M155 B110 ; n
M155 S1   ; Send the current buffer
```

As a side effect it also supports:
- `M155 B106 B98 B114 B97 B122 B105 B111`
- `M155 B 106 B 98 B 114 B 97 B 122 B 105 B 111`
- `M155 B106B98B114B97B122B105B111`

**Comment by thinkyhead** [comment]
> I'm not pushing nothing outside the envelope here…. As a side effect….

The ability to parse multiple `B` arguments isn't a side-effect, but an added feature, requiring the handler to have too much "inside knowledge" about the original GCode string input. The other handlers abstract this stuff, so that we can change the parser or data storage method without changing the handlers:

``` cpp
if (code_seen('B')) my_val = code_value_short();
float my_flot = code_seen('B') ? code_value() : 0.0;
```

The code you've posted adds low-level parsing into the middle of a handler instead of using the designated accessors. It can't be merged in its current form.

Sorry if I seem to be all over the map about this, but we don't have the luxury to treat this as our own personal project, so I have to let it sink in how such changes affect the larger ecosystem.

**Comment by jbrazio** [comment]
> Sorry if I seem to be all over the map about this

No worries.

```
SENDING:M155 A63
twibus::sendto: 63
SENDING:M155 B106 B98 B114 B97 B122 B105 B111
twibus::addbyte: 106
SENDING:M155 S1
twibus::send()
```

## ds1z_quickprint1

```
SENDING:M155 A63
twibus::sendto: 63
SENDING:M155 B77
twibus::addbyte: 77
SENDING:M155 B97
twibus::addbyte: 97
SENDING:M155 B114
twibus::addbyte: 114
SENDING:M155 B108
twibus::addbyte: 108
SENDING:M155 B105
twibus::addbyte: 105
SENDING:M155 B110
twibus::addbyte: 110
SENDING:M155 S1
twibus::send()
```

## ds1z_quickprint2

```
SENDING:M156 A63 B5
twibus::sendto: 63
twibus::reqbytes(): 5
echo:I2C-reply: from:63 bytes:5 data:hello
```

## ds1z_quickprint3

@thinkyhead is there a reason why `DEBUG_INFO` is disabled ?

``` cpp
// FOR MOMENT NOT ACTIVE
//if (marlin_debug_flags & DEBUG_INFO) SERIAL_ECHOLNPGM(MSG_DEBUG_INFO);
```

**Comment by thinkyhead** [comment]
I'm not sure why `DEBUG_INFO` is disabled. It's not a very good name for a flag, though, is it? I don't even remember what "info" it's about.

**Comment by jbrazio** [comment]
I believe who wrote the code was thinking about the standard syslog severity levels where you have warning, notice, info, debug. 

The flag is currently unused on the code, there are just only placeholders to allow it's activation trough `M111`. I tried to use `DEBUG_ECHO` but it is unusable because everything your send is echoed back; do you mind if I activate `DEBUG_INFO` ?

**Comment by thinkyhead** [comment]
I don't mind if it's re-enabled, since all debugging is turned off by default. It would be good if you researched to find out why it was disabled in the first place, for peace of mind. Check the Blame and History for those lines.

**Comment by jbrazio** [comment]
It was introduced by 76306f9073205390a4da0ba7204bf37fc2ae36d1 but the initial commit was already commented, so it was never used.

**Comment by jbrazio** [comment]
> since all debugging is turned off by default

I have been digging deeper and this is not true, in fact there is an inconsistency on the code.
I will submit another PR so we do not pollute this one.

File: Marlin/twibus.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+
+    // Protect against buffer overflows if the number of received bytes
+    // is less than the number of requested bytes
+    uint8_t wba = Wire.available();
+    for (int i = 0; i < wba; i++) SERIAL_CHAR(Wire.read());
+    SERIAL_EOL;
+
+    // Reset the buffer after sending the data
+    this->reset();
+}
```
Comment (thinkyhead) [inline Marlin/twibus.cpp]: Indentation should be two spaces in these `twibus.*` files.

If you add `EXPERIMENTAL_I2CBUS` to either (all) `Configuration.h` or `Configuration_adv.h` files, then I don't mind merging this so users can mess with it (once they know about it, of course).

**Comment by thinkyhead** [comment]
> this is not true

Ah yes, I found that out quickly too. I didn't have ready enough access to the interwebs to post a correction – sorry about that! The good thing about limited net connectivity is I get to do a lot more testing on the RepRap before I post new code.

**Comment by jbrazio** [comment]
@thinkyhead please review, should be ready for merge.

**Comment by thinkyhead** [comment]
Looks good! I did a little patch-up on the class name (`TWIBus`), plus I wrapped the `twibus.cpp` code to keep it from compiling if the option isn't enabled. Since the feature is disabled by default, it's safe to merge, so I will merge it pretty soon.

**Pull request closed without merging**